client-go/examples: Update CRUD Deployment sample

nulltrope opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**: 
PR motivated by [#128](https://github.com/kubernetes/client-go/issues/128), namely updating the CRUD example with the following:
- Add new step which demonstrates rolling back deployments 
- Cleanup retry loop for `Update()` steps
- Make `-kubeconfig` flag optional when running example (same as out-of-cluster example)
- Update `README.md` to reflect changes 

**Special notes for your reviewer**:
My first Kubernetes contribution- feedback very welcome!

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

/cc @ahmetb @caesarxuchao

Commit:
client-go: Update CRUD example

This updates the create-update-delete-deployment example with the following:
Add rollback step to demonstrate rolling back deployments with client-go.
Modify the for-loops used in both Update steps to Get() the latest version
of the Deployment from the server before attempting Update().
This is necessary because the object returned by Create() does
not have the new resourceVersion, causing the initial Update() to always fail
due to conflicting resource versions. Putting the Get() at the top of the
loop seems to fix this bug.
Make -kubeconfig flag optional if config is in default location, using the
same method found in the out-of-cluster example.

Patch is motivated by effort to improve client-go examples.

Signed-off-by: John Kelly <jekohk@gmail.com>

ahmetb commented on staging/src/k8s.io/client-go/examples/create-update-delete-deployment/main.go:
```diff
@@ -100,32 +107,63 @@ func main() {
+		if _, err = deploymentsClient.Update(result); err != nil {
+			if errors.IsConflict(err) {
+				// Deployment was modified since last retrieved, need to retry
+				fmt.Println("Conflicting resource versions, retrying")
```
Conflicting resource versions, retrying &rarr; Deployment was modified since last retrieved, retrying with latest

ahmetb commented on staging/src/k8s.io/client-go/examples/create-update-delete-deployment/main.go:
```diff
@@ -61,6 +65,9 @@ func main() {
 					Labels: map[string]string{
 						"app": "demo",
 					},
+					Annotations: map[string]string{
```
not sure why you needed to add an annotation. :) this example was already designed to change replica count (visible in kubectl get/describe, too)

caesarxuchao replied:
I think it's to demonstrate the rollback? Annotations is part of the podTemplate, changing templates results in a new revision, so that the deployment can be rolled back.

jekohk replied:
mainly to make it more obvious that the rollback occurred- replica count will not change after rollback. But you're right, it's not really necessary, if it causes confusion I can remove it :)

ahmetb replied:
> so that the deployment can be rolled back.

I think the replicaCount also creates a new Deployment revision even though it's not part of PodTemplate?

But yeah it's good to demonstrate pod rollout as well, but right now adding a `fizz=buzz` annotation and not explaining doesn't make it clear to the reader why we're updating two things.

jekohk replied:
As far as I know, only changing replicaCount will not trigger a rollout
> Note: A Deployment’s revision is created when a Deployment’s rollout is triggered. This means that the new revision is created if and only if the Deployment’s pod template (.spec.template) is changed, for example if you update the labels or container images of the template.

[See here](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)

caesarxuchao replied:
Instead of changing annotations, how about changing to a newer/older version of the [nginx](https://github.com/jekohk/kubernetes/blob/ccfdd2ae7bfebdc804da6fb30875906e97ac106d/staging/src/k8s.io/client-go/examples/create-update-delete-deployment/main.go#L76) image? That's a more realistic use case as well.

jekohk replied:
agreed, changing the image is a more realistic use case & better illustrates how rollbacks work.

ahmetb replied:
+1

ahmetb commented:
/lgtm

I think without the `// Retrieve latest version of Deployment before modifying and updating` part the current example is buggy. 

@caesarxuchao to approve

ahmetb commented:
/assign @caesarxuchao
to approve

caesarxuchao commented on staging/src/k8s.io/client-go/examples/create-update-delete-deployment/main.go:
```diff
@@ -100,32 +107,63 @@ func main() {
 
 	for {
+		// Retrieve latest version of Deployment before modifying and updating
+		result, err = deploymentsClient.Get("demo-deployment", metav1.GetOptions{})
```
Consider using the retry utility function: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/retry/util.go#L61

jekohk replied:
Yup it was just brought to my attention that client-go has this utility function, definitely a better way to handle this.. will update with that change

ahmetb replied:
I wasn't aware of RetryOnConflict while developing this sample. It looks dope. 👌

caesarxuchao commented on staging/src/k8s.io/client-go/examples/create-update-delete-deployment/main.go:
```diff
@@ -100,32 +107,63 @@ func main() {
+	fmt.Println("Rolling back deployment...")
+
+	// Use same method as above to avoid version conflicts
+	for {
```
Same here, consider using the retry utility function.

Commit:
client-go: use retry util in CRUD example

This updates the create-update-delete-deployment example with the following:
Make use of client-go retry util in Update() steps instead of simple for loops.
Using RetryOnConflict is generally better practice as it won't become stuck in a retry loop and uses exponential backoff to prevent exhausting the apiserver.
Instead of changing annotations to demonstrate Updates/Rollbacks, change the container image as it is less confusing for readers and a better real-world example.
Improve comments and README to reflect above changes.

Signed-off-by: John Kelly <jekohk@gmail.com>

jekohk commented:
before I commit the new changes, is there anything extra I need to do since I am introducing a new dependency k8s.io/client-go/util/retry to the program?

I noticed that `BUILD` has a list of deps but not sure if it needs to be manually edited or not

ahmetb commented:
@jekohk bazel checks seem successful, I wouldn't worry about it.

jekohk commented:
@ahmetb `update-bazel` script added a line to the `BUILD` file, so I included it in my latest commit. Let me know if you have any other suggestions :)

ahmetb commented on staging/src/k8s.io/client-go/examples/create-update-delete-deployment/main.go:
```diff
@@ -97,35 +101,52 @@ func main() {
+	fmt.Println("Rolling back deployment...")
+	// Once again use RetryOnConflict to avoid update conflicts
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
+		result, err = deploymentsClient.Get("demo-deployment", metav1.GetOptions{})
```
this err, the err above on the left-side, and the err above in the anonymous function signature, is prone to scope issues. let's try to name them separately to avoid confusion

jekohk replied:
good point, I should've caught that- I used the retry example without thinking.

will address that and do more thorough local testing, thanks for pointing that out!

Commit:
client-go: fix err scope in CRUD example

This fixes some scope isses that were introduced by shadowing vars inside anonymous functions as well as using a naked return.
Fixed by using unique err names and explicitly returning errors.
Additional improvement is using the HomeDir() util function provided by client-go instead of including a helper function at the bottom of this example.

Signed-off-by: John Kelly <jekohk@gmail.com>

ahmetb commented:
/ok-to-test
/lgtm
/assign @caesarxuchao

ahmetb commented:
ping @caesarxuchao.

jekohk commented:
@ahmetb @caesarxuchao is another fetch/rebase needed so this passes tests again?

ahmetb commented:
these look like flakes.
let's give it another try, if it doesn't work again we can try rebasing.

/retest

ahmetb commented:
Try rebasing we'll approve again. Or @caesarxuchao can merge directly maybe since these are likely flakes.

caesarxuchao commented:
/lgtm
